nncf/experimental/common/pruning/nodes_grouping.py

```diff
@@ -0,0 +1,428 @@
+                       prune_operations_types) -> List[PruningGroup]:
+    """
+    Determines how nodes of the given types should be pruned: which nodes should be pruned together, along which
+    dimension, how many sequent channels with which offset. It's done by initializing PropagationMask's on the
+    operations with prunable paramaters (producers of pruning) and propagating them through the execution graph.
+
+    :param graph: nncf graph to initialize and propagate masks.
+    :param pruning_operations_metatypes: registry with operation metatypes pruning algorithm is aware of, i.e.
+    metatypes describing operations with common pruning mask application and propagation properties, e.g.
+    IdentityMaskForwardOps unifies operations that propagate pruning masks as is (relu, swish etc.), whereas
+    Convolution unifies different convolution operations (conv1d, conv2d, conv3d) which accepts some input masks and
+    provide some output masks.
+    :param prune_operations_types: types of operations with prunable parameters.
+    :return: list of groups with parameters of pruning.
+    """
+    # 1. Initialize masks for producing nodes
+    all_nodes_to_prune = graph.get_nodes_by_types(prune_operations_types)  # type: List[NNCFNode]
+    roots = {}
+    for node in all_nodes_to_prune:
+        assert isinstance(node.layer_attributes, (LinearLayerAttributes, ConvolutionLayerAttributes))
+        pruning_dim = node.layer_attributes.get_target_dim_for_compression()
+        output_tensors_shapes = [x.tensor_shape for x in graph.get_output_edges(node)]
+        assert len(output_tensors_shapes) == 1 or len( set(output_tensors_shapes)) <= 1, node.node_name
+        output_tensors_shape = output_tensors_shapes[0]
+        target_output_dim_for_compression = len(output_tensors_shape) - 1
+        root_group = PropagationGroup(
+            blocks=[
+                PropagationBlock(
+                    producer=MaskProducer(node.node_id),
+                    pruning_dimension=pruning_dim
+                )
+            ]
+        )
+        mask = PropagationMask(
+            dim_groups_map={
+                target_output_dim_for_compression: [root_group]
+            }
+        )
+        roots[node.node_id] = root_group
+        node.data['output_mask'] = mask
+
+    def get_attributes_fn(node: NNCFNode) -> Dict[str, Any]:
+        result = {'metatype': str(node.metatype.name), 'node_id': str(node.node_id)}
+        if node.layer_attributes:
+            result.update(map(lambda pair: (pair[0], str(pair[1])), node.layer_attributes.__dict__.items()))
+        if 'output_mask' in node.data:
+            output_mask = node.data['output_mask']
+            if output_mask:
+                result['output_mask'] = str(output_mask)
+        return result
+
+    try:
+        # 2. Propagate masks
+        MaskPropagationAlgorithm(
+            graph, pruning_operations_metatypes).mask_propagation()
+    finally:
```

Comment:
I know this experimental, but still, is it ok to commit the code that will save debug data every time this function is called?

Reply:
It happens only once for the model.
I'd move this under `is_debug()` condition when the algorithm will be more mature. Otherwise, it's not convenient to debug it.